### DISCUSSION TITLE
Help adding "@babel/plugin-proposal-pipeline-operator"

### DISCUSSION THREAD

**Discussion by bbergeron-xlkey** [vercel/next.js#73527] [Help] [answered]
### Summary

I really want a pipeline operator in JS, which I should be able to get if I use "@babel/plugin-proposal-pipeline-operator". However I have no idea how to set this up, and neither the documentation, Google, or ChatGPT could come up with a coherent answer that yielded any results. Can someone help me?

**Comment by icyJoseph** [comment, marked as answer]
- https://nextjs.org/docs/pages/building-your-application/configuring/babel

Create `.babelrc`:

```json
{
  "presets": ["next/babel"],
  "plugins": [
    [
      "@babel/plugin-proposal-pipeline-operator",
      {
        "proposal": "minimal"
      }
    ]
  ]
}
```

Install packages:

```json
    "@babel/core": "^7.26.0",
    "@babel/plugin-proposal-pipeline-operator": "^7.25.9",
    "@babel/runtime": "^7.26.0",
```

And make sure, you don't use `next/font`: https://nextjs.org/docs/messages/babel-font-loader-conflict

Be careful with depending too much on this feature though. I has gone through many changes and proposals, and it has been stuck at stage 2, for a while.

**Reply by bbergeron-xlkey** [reply]
Thanks, Joseph. Unfortunately, even a basic .babelrc gives me a bunch of compilation errors, and my language server still doesn't recognize |> as a valid operator. Suppose I figure out how to make Babel work with Next.js—do you know what the next step would be to get the pipeline operator working with the language server (ESLint)?

**Reply by icyJoseph** [reply]
Odd, I got it to work with Next.js ~ perhaps your project needs more plugins? 

Start a new project and make it work there first. It took me 5 minutes to set up.

What error do you have?

Yeah the LSP and such, idk yet.. this feature is stage 2, with many flavors, and many tools only incorporate stage 3+

**Reply by bbergeron-xlkey** [reply]
I'm getting this:

```
  x Server actions must be async functions
     ,-[/home/bbergeron/Projects/XLKey/xlhub-platform/feature/dropbox/xlhub-frontend/src/app/(dashboard)/televersements/actions.ts:97:1]
  97 |     return _ref3.apply(this, arguments);
  98 |   };
  99 | }();
 100 | export function updateFileAction(_x3) {
     :                 ^^^^^^^^^^^^^^^^
 101 |   return _updateFileAction.apply(this, arguments);
 102 | }
 103 | function _updateFileAction() {
     `----
```

Here's the source code:

```ts
export async function updateFileAction(form: FormData) {
	return pipe(
		TE.Do,
		TE.bind("config", getFileUploadConfig),
		TE.bind("identity", getCookiesIdentity),
		TE.flatMap(({ config, identity }) => {
			const { farmId } = identity;

			return pipe(
				form,
				validateUpdateFilePayload,
				V.match(TE.right, e => TE.left(e)),
				TE.flatMap(payload => updateFile(config)({
					farmId,
					...payload
				})),
				TE.map(() => revalidatePath("/"))
			)
		}),
		makeServerActionResponse
	)
}
```

Removing `.babelrc` fixes the issue, so Babel definitely has something to do with it. As you can see, I'm using fp-ts's `pipe` function throughout my code as a surrogate for a pipeline operator, but this gives me poor error analysis with LSP. Hence, I'm looking for a real operator. Unless I can get the pipeline operator to work with LSP, I'd better stick with `pipe` for now.

**Reply by icyJoseph** [reply]
Right, looks like a known issue, https://github.com/vercel/next.js/issues/57966, probably because of how async/await is transpiled, by Babel. That GitHub issue has no solution I could see, but this Stack overflow question has at least a suggestion, https://stackoverflow.com/a/77060062, could you try that?

I'm not sure if you are going in the right direction here. Formally speaking, JavaScript has no pipe operator whatsoever, as of writing in 2024. What we have is a stage 2 proposal, with several variations. I remember trying it out as far back as 2018, but back then we didn't use it in production, and today I won't even bother to try it out for fun.

Tooling would have to commit a lot of time to support variations of the proposal, and adjust accordingly when it gets, if ever, to stage 3.

Maybe people who have more experience with LSPs and such can hack it, but is it even worth it? 

Perhaps somewhat related, https://jrsinclair.com/articles/2022/what-if-the-team-hates-my-functional-code/

**Reply by icyJoseph** [reply]
I'll wrap up saying that, to the original question, an answer, verified, has been given, although there's a known issue with server actions and Babel transpiled code.

**Reply by bbergeron-xlkey** [reply]
I appreciate an opinionated answer, but though it’s worth considering, I must respectfully assert that’s not what I’m asking for here. Nonetheless, you’ve helped me a great deal, and the question is indeed answered. I’m marking it as resolved. Thanks again, Joseph!
